dev_setup.sh

```diff
@@ -0,0 +1,113 @@
+#  * macOS 14 (AArch64)
+#
+# This script installs everything needed to get a working development
+# environment for cryptol. Any new environment requirements should be
+# added to this script.
+#
+
+set -e
+
+HERE=$(cd `dirname $0`; pwd)
+LOG=$HERE/dev_setup.log
+
+function notice {
+    echo "[NOTICE] $*"
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        mkdir -p `dirname $LOG`
+        echo $* >>$LOG
+        if ! $* >>$LOG 2>&1
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $HERE
+    notice "Updating submodules"
+    git submodule update --init
+}
+
+function install_ghcup {
+    if ! ghcup --version &> /dev/null
+    then
+        notice "Installing ghcup, GHC, and cabal"
+        # technically the installation only requires cabal, but it's 
+        # recommended to get the whole GCH shebang in one package
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
+    else
+        notice "Using existing ghcup installation"
+    fi
+
+}
+
+# Indicate whether this is running macOS on the Apple M series hardware.
+function is_macos_aarch {
+    # Is it running macOS?
+    if [[ "$OSTYPE" != "darwin"* ]]; then
+        notice "dev setup does not currently support your OS / hardware"
+        exit 1 
+    fi
+
+    # Does it use an M-series chip?
+    chip=$(system_profiler SPHardwareDataType | grep Chip)
+    [[ $chip == *M1* || $chip == *M2* || $chip == *M3* ]]
+}
+
+function install_gmp {
+    if is_macos_aarch; then
+        notice "Installing gmp via Homebrew, if it's not already installed"
+        logged brew list gmp || brew install gmp
+
+        # on macOS 12 (x86_64), I think homebrew uses different locations
+        notice "You may need to add the following environment variables to your '.profile': \n"\
+            "export CPATH=/opt/homebrew/include\n"\
+            "export LIBRARY_PATH=/opt/homebrew/lib\n"
+    fi
+}
+
+function install_what4_solvers {
+    if ! cvc4 --version &> /dev/null || ! cvc5 --version &> /dev/null; then
+        notice "Installing cvc4 and/or cvc5 solvers"
+
+        # There are different URLs for other supported platforms
+        if is_macos_aarch; then
+            what4_solvers_url="https://github.com/GaloisInc/what4-solvers/releases/download/snapshot-20240212/macos-14-ARM64-bin.zip"
+        fi
+
+        solvers_dir=$(mktemp -d)
+        curl --proto '=https' --tlsv1.2 -sSfL $what4_solvers_url > "$solvers_dir/solvers.bin.zip"
+        cd $solvers_dir
+        logged unzip solvers.bin.zip
+        rm solvers.bin.zip
+        
+        # If we want to install more solvers by default, we can do so here,
+        # although we might need a different check than `--version`
+        for solver in cvc4 cvc5; do
```

Comment:
Reading further, I guess that suggestion doesn't make sense for this second part where you are installing *all* supported CVCs.

Reply:
Yeah, this feels a bit awkward; I was originally going to install all the solvers in the repo but they all have different commands to check if they're installed (can't just do `$solver --version`) and these are the only two needed to run the tests. I figured if users want other specific solvers they can double check the script and do the equivalent thing for themselves.

Reply:
Oh, interesting. So the solver installation here is only about getting the tests running? I ask because I could imagine that for getting an *installation* of cryptol running, a more robust solver installation process might be needed but this could be left as-is.

Reply:
One thought on development setup vs. installation is that I think that unifying how the solvers are installed and managed across both of those would be really important for ensuring that what we test on, what the user uses, and what we say we support, are all ensured to be overlapping. So that's more food for thought as other infrastructure gets touched here, since ISTM it would be really nice if there was a single source of truth for solver management that is easy to update, and then we can be confident that CI, development, and deployment environments are all using exactly the same things.

Reply:
Yeah, I agree that it would be nice to push any interesting logic here into the what4_solvers repo to let it be called anywhere. I don't know how people actually use cryptol to say whether it makes sense to install one or more of them, which is why I defaulted to just getting the tests to run.